Changelog 178 (#161)
intellij-rust/intellij-rust.github.io: _posts/2022-09-12-changelog-178.markdown

---
layout: post
title: "IntelliJ Rust Changelog #178"
date: 2022-09-12 13:00:00 +0300
---

## New Features

* [#9268], [#9259] Complete proc macros in attributes

  {% include gif-img.html path="/assets/posts/changelog-178/complete_attr_proc_macros" w="350px" %}

* [#9251] Detect [E0728](https://doc.rust-lang.org/error-index.html#E0728) compiler error

  {% include gif-img.html path="/assets/posts/changelog-178/await_outside_async" w="350px" %}

> * [#9272] Detect [E0434](https://doc.rust-lang.org/error-index.html#E0434) compiler error

Reviewer: And maybe use the following format
```
- Detect new compiler errors:

  - %Short error description% ([%error_code%])
  ```

  {% include gif-img.html path="/assets/posts/changelog-178/use_outer_variable" w="350px" %}

* [#9209] Provide debugger pretty-printers to render range types (e.g. `core::ops::range::Range`)

  <img src="/assets/posts/changelog-178/pretty_printers_range_types.png" width="700px"/>

* [#8961] Take into account build target during collection information about the project model and don't download and analyze [platform-specific](https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html#platform-specific-dependencies) dependencies not suitable for the current platform

> * [#9256] Detect [E0733](https://doc.rust-lang.org/error-index.html#E0733) compiler error

Reviewer: Let's mention that we still don't detect complex cases like described in https://github.com/intellij-rust/intellij-rust/issues/9319

  {% include gif-img.html path="/assets/posts/changelog-178/recursive_async_function" w="350px" %}

* [#9269] Add completions for new Clippy and Rustc lints (by [@afetisov])

## Performance Improvements

> * [#9236] Enable again [the optimization](https://github.com/intellij-rust/intellij-rust/pull/9021) reverted in the previous release. Slightly speed up name resolution

Reviewer: ```suggestion
* [#9236] Enable again name resolution [optimization](https://github.com/intellij-rust/intellij-rust/pull/9021) reverted in the previous release
```

* [#9171] Slightly speed up name resolution

> * [#9255] Speed up "Optimize import" action when "Unused imports" inspection is enabled

Reviewer: ```suggestion
* [#9255] Speed up `Optimize import` action when `Unused imports` inspection is enabled
```

## Fixes

* [#9235] Fix broken `anyhow` compilation when `org.rust.cargo.evaluate.build.scripts` [experimental feature](https://plugins.jetbrains.com/plugin/8182-rust/docs/rust-faq.html#experimental-features) is enabled

* [#9202] Fix resolve of items inside module declared in `include!`-ed file

* [#9295] Fix search of the plugin settings in `Find Action` dialog and `Settings` window which was broken in [#8709](https://github.com/intellij-rust/intellij-rust/pull/8709)

* [#9238] Allow leading `|` in nested or-patterns (by [@afetisov])

* [#9288] Always reload the project model after changes in settings which affect the model

* [#9267] Fix auto import of attr proc macro with the same name as crate

  {% include gif-img.html path="/assets/posts/changelog-178/import_attr_proc_macro" w="350px" %}

> * [#9280] Fix expansion of local (located inside function bodies) attr/derive procedural macros

Reviewer: ```suggestion
* [#9280] Fix expansion of local (located inside function bodies) [attribute](https://doc.rust-lang.org/reference/procedural-macros.html#attribute-macros) and [derive](https://doc.rust-lang.org/reference/procedural-macros.html#derive-macros) procedural macros
```

* [#9205] Update debugger pretty-printers to render `CString` and `CStr` properly

* [#9233] Fix cargo feature state inference in several cases

* [#9265] Fix auto-import variants for `error::Error` path on nightly toolchain

* [#9225] Properly parse environment variables in [cargo config](https://doc.rust-lang.org/cargo/reference/config.html#env)

* [#9234] Show error message in Sync tool window if project directory does not exist

> * [#9270] Simplify "Inline function" refactoring dialog

Reviewer: ```suggestion
* [#9270] Simplify `Inline function` refactoring dialog
```

Reviewer: Let's specify what exactly was changed

---
layout: post
title: "IntelliJ Rust Changelog #178"
date: 2022-09-12 13:00:00 +0300
---

## New Features

* Detect new compiler errors:

  * [#9251] `await` outside `async` function ([E0728](https://doc.rust-lang.org/error-index.html#E0728))

    {% include gif-img.html path="/assets/posts/changelog-178/await_outside_async" w="350px" %}

  * [#9272] Use of variable from outer function ([E0434](https://doc.rust-lang.org/error-index.html#E0434))

    {% include gif-img.html path="/assets/posts/changelog-178/use_outer_variable" w="350px" %}

>   * [#9256] An async function used recursion without boxing ([E0733](https://doc.rust-lang.org/error-index.html#E0733)). Note that detection of indirect async recursion (two async functions calling each other) is not supported now

Reviewer: ```suggestion
  * [#9256] An async function used recursion without boxing ([E0733](https://doc.rust-lang.org/error-index.html#E0733)). Note, detection of indirect async recursion (two async functions calling each other) is not supported for now
```

* [#9209] Provide debugger pretty-printers to render range types (e.g. `core::ops::range::Range`)

  <img src="/assets/posts/changelog-178/pretty_printers_range_types.png" width="700px"/>

* [#8961] Take into account build target during collection information about the project model and don't download and analyze [platform-specific](https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html#platform-specific-dependencies) dependencies not suitable for the current platform

* [#9269] Add completions for new Clippy and Rustc lints inside lint attributes like `allow`, `warn`, etc. (by [@afetisov])

## Performance Improvements

* [#9236] Enable again name resolution [optimization](https://github.com/intellij-rust/intellij-rust/pull/9021) reverted in the previous release

* [#9171] Slightly speed up name resolution

* [#9255] Speed up `Optimize import` action when `Unused imports` inspection is enabled

## Fixes

* [#9235] Fix broken `anyhow` compilation when `org.rust.cargo.evaluate.build.scripts` [experimental feature](https://plugins.jetbrains.com/plugin/8182-rust/docs/rust-faq.html#experimental-features) is enabled

> * [#9202] Fix resolve of items inside module declared in `include!`-ed file

Reviewer: ```suggestion
* [#9202] Fix name resolution of items inside module declared in `include!`-ed file
```

* [#9295] Fix search of the plugin settings in `Find Action` dialog and `Settings` window which was broken in [#8709](https://github.com/intellij-rust/intellij-rust/pull/8709)

* [#9238] Allow leading `|` in nested or-patterns (by [@afetisov])

* [#9288] Always reload the project model after changes in settings which affect the model

> * [#9267] Fix auto import of attr proc macro with the same name as crate

Reviewer: ```suggestion
* [#9267] Fix auto-import of attribute procedural macro with the same name as crate
```

  {% include gif-img.html path="/assets/posts/changelog-178/import_attr_proc_macro" w="350px" %}

* [#9280] Fix expansion of local (located inside function bodies) [attribute](https://doc.rust-lang.org/reference/procedural-macros.html#attribute-macros) and [derive](https://doc.rust-lang.org/reference/procedural-macros.html#derive-macros) procedural macros

* [#9205] Update debugger pretty-printers to render `CString` and `CStr` properly

* [#9233] Fix cargo feature state inference in several cases

* [#9265] Fix auto-import variants for `error::Error` path on nightly toolchain

* [#9225] Properly parse environment variables in [cargo config](https://doc.rust-lang.org/cargo/reference/config.html#env)

> * [#9234] Show error message in Sync tool window if project directory does not exist

Reviewer: ```suggestion
* [#9234] Show error message in `Sync` tool window if project directory does not exist
```

> * [#9270] Simplify `Inline function` refactoring dialog (remove unused options `Search in comments`)

Reviewer: ```suggestion
* [#9270] Simplify [`Inline function`](https://plugins.jetbrains.com/plugin/8182-rust/docs/rust-refactorings.html#extractmethod-refactoring) refactoring dialog (remove unused options `Search in comments`)
```
